[feat] QnaCard 컴포넌트 구성

oooppq opened the pull request:
## PR 요약

> QnaCard component와 이 component에 필요한 child component들 완성

## 변경된 점

> QnaCard components 추가, types/utils 폴더 추가

## 이슈 번호

> X

oooppq committed:
chore: storybook static directory 설정

Image 컴포넌트에서 src의 경로가 바르게 인식되도록 static directory를 public 폴더로 설정해주었다.

oooppq committed:
feat: 필요한 아이콘 추가 & qna type 정의

oooppq committed:
feat: qnaCard 컴포넌트 구성

oooppq committed:
feat: 반응 컴포넌트 구현

각각의 반응이 1 이상이면 곧바로 display 되도록 했고, 반응 펼치기 버튼을 누르면 모든 반응 버튼을 볼 수 있도록 구성했다.
구성하면서 내용이 많아져 footer 컴포넌트를 새롭게 정의하여 분리했다.

oooppq committed:
refactor: 마이너한 변경사항 적용

home directory 내부에 qnaCard 폴더를 추가하여 다른 컴포넌트들과 구분했고,
test 용으로 작성했던 console.log를 모두 제거했으며,
login button component의 click handler prop name을 수정했다.

yeon0914 commented on components/home/qnaCard/PostTemplate.tsx:
```diff
@@ -0,0 +1,47 @@
+
+// import { useRouter } from 'next/navigation';
+import { convertDate } from '@/utils';
+import PostTemplateView from './PostTemplateView';
```
저희 절대경로 사용하기로 했던거같은데 맞나요..?

oooppq replied:
반영하겠습니다~

yeon0914 commented on components/home/qnaCard/PostTemplate.tsx:
```diff
@@ -0,0 +1,47 @@
+
+  return (
+    <PostTemplateView
+      postInfo={{ ...postInfo, date: convertDate(postInfo.date) }}
```
이렇게 props를 새로운 객체를 생성해서 자식컴포넌트로 전달하는 방식은 불필요한 리렌더링을 유발하기 때문에 convertDate는 자식 컴포넌트에서 처리하는게 더 좋을것 같습니다~!

oooppq replied:
넵 수정하겠습니다~~

yeon0914 commented on components/home/qnaCard/PostTemplate.tsx:
```diff
@@ -0,0 +1,47 @@
+}: PostTemplateProps) => {
+  // const router = useRouter();
+
+  const handleClickDMButton: React.MouseEventHandler = (e) => {
```
함수 이벤트 타입 지정방식에 있어서
1. 대균님이 하신것처럼 const handleClickDMButton: React.MouseEventHandler = (e) => {} 로 지정하는 방법
2. const handleClickDMButton = (e: React.MouseEvent<HTMLElement>) => {} 로 지정하는 방법 이 있는데

이거도 어떻게 할지 정하면 좋을거같아요 저는 주로 2번 방식대로 사용해서요! 어떻게 생각하시나요?

oooppq replied:
원래 2번쓰시던거면 2번 방식으로 바꿀게요! 저는 둘 다 상관없습니다

yeon0914 replied:
저는 2번방식이 코드가 직관적이어서 더 선호해요

yeon0914 commented on components/home/qnaCard/PostTemplateView.tsx:
```diff
@@ -0,0 +1,63 @@
+    date: string;
+  };
+  isQuestion: boolean;
+  onClickDMButton?: React.MouseEventHandler;
```
이 함수들은 왜 옵셔널인가용?

oooppq replied:
혹시 저 컴포넌트가 다른 곳에서 사용될 때 DM버튼이 필요 없는 경우가 있을 것 같아서 옵셔널로 해놨습니다!

yeon0914 commented on components/home/qnaCard/QnaCardView.tsx:
```diff
@@ -0,0 +1,70 @@
+  onClickRejectButton?: () => void;
+}
+
+const QnaCardView = ({
```
답변이 완료된 QnaCard와 답변 대기중인 QnaCard를 따로 분리를 하는건 어떨까라는 생각이 드는데 어떠신가용? 
뭔가 기능이 더 들어가면 하나의 컴포넌트가 너무 커질거같아가지고...

oooppq replied:
새질문, 답변완료, 거절질문 다 하나로 하려고 했는데 그냥 따로따로 컴포넌트 분리할게요!

yeon0914 replied:
넹 질문 타입이 총4가지니깐 대균님이 적절히 나누시면 될거같아용~

yeon0914 commented on utils/index.ts:
```diff
@@ -0,0 +1,27 @@
+// eslint-disable-next-line import/prefer-default-export
+export const convertDate = (postedDate: Date) => {
```
오 이거 유틸 추가해주셨네용 👍
위에 eslint 주석은 왜 들어간건가요?
그리고 네이밍을 기능이 조금더 잘 드러나도록 변경해주시면 좋을거같습니다!

oooppq replied:
저거 한 파일에서 export 하나만 할거면 export default ~~ 로 선언하라고 에러띄워서 일단 막아놨어요! 다른 util 추가되면 지우면 될 것 같아요

yeon0914 commented on types/index.ts:
```diff
@@ -0,0 +1,23 @@
+export interface TReactions {
```
like, thumb, check 같은 것들을.. 상수화하면 좋을거같은데 아래와 같은 방식은 어떤가용?
```
export const Reaction = {
  LIKE: 'LIKE',
  THUMB: 'THUMB',
  CHECK: 'CHECK',
  SMILE: 'SMILE',
  SAD: 'SAD'
} as const;

export type TReactions = {
  [key in keyof typeof reaction]: number;
};
```

oooppq replied:
저거 백엔드에서 어떤 형식으로 만들어줄지 모르겠어서 일단 해놨던건데, 말씀해주신대로 바꿔놓고 백엔드에 형식 맞춰달라고 말씀을 드릴까요??

yeon0914 replied:
넹 좋죠

oooppq committed:
refactor: 코드 리뷰 반영

yeon0914 commented on components/home/PostTemplate.tsx:
```diff
@@ -0,0 +1,46 @@
+    date: Date;
+  };
+  isQuestion: boolean;
+  containerStyles?: string;
```
앗 이거를 놓쳤는데 containerStyles를 넘겨주는거보다는 style 관련은 확실히 View로 분리되는게 좋을거같아요!
해당 style 분기 조건을 넘겨주고, style은 View 내부에서 const 로 선언후에 관리하는게 어떨까용?
저도 style 분기가 있었는데 이런식으로 작성했습니다!

https://github.com/Cling-Glee/glee-front/blob/4b44644bbd6129bc5138ea4b61ff83d58ff0c18b/components/home/tab/TabView.tsx#L18

oooppq replied:
앗넵 디자인 변경사항하고 적용해서 커밋하나만 더 올릴게요~~

yeon0914 reviewed (approved):
수고하셨습니다!!👍 마지막 코멘트만 확인하시고 머지하시죵!

oooppq committed:
refactor: 컴포넌트 간 props 구성 변경

card를 qna 카드와 newQuestion 카드로 분리하면서, 반복되는 코드가 생겼기 때문에
dm button click handler를 postTemplate 컴포넌트에서 선언하도록 했고, dm버튼 필요 유무를
prop으로 받도록 하여, view에서 해당 flag를 기준으로 dm버튼 display 유무를 결정하도록 했다.

oooppq committed:
feat: accepted state의 질문 구현

oooppq committed:
refactor: reaction 애니메이션 수정
